File: tests/oracle/comment_warning_codec_1204.cc
/*
 * COMMENT-WARNING-CODEC-0001: locked Ghidra 12.0.4 oracle for
 * Comment::{encode,decode} and CommentDatabaseInternal's ordered in-memory
 * add/deduplicate/filter/codec behavior.
 */

#include <iostream>
#include <map>
#include <sstream>
#include <string>
#include <vector>

#include "comment.hh"
#include "marshal.hh"
#include "translate.hh"

using namespace ghidra;

class FixtureTranslate final : public Translate {
  VarnodeData dummy_register;

public:
  FixtureTranslate() {
    setBigEndian(false);
    setUniqueBase(0);
    insertSpace(new ConstantSpace(this, this));
    insertSpace(new AddrSpace(this, this, IPTR_PROCESSOR, "other", false, 8, 1, 1,
                              AddrSpace::hasphysical, 0, 0));
    insertSpace(new UniqueSpace(this, this, 2, 0));
    insertSpace(new AddrSpace(this, this, IPTR_PROCESSOR, "ram", false, 8, 1, 3,
                              AddrSpace::hasphysical, 0, 0));
    setDefaultCodeSpace(3);
    dummy_register.space = getSpace(3);
    dummy_register.offset = 0;
    dummy_register.size = 8;
  }

  void initialize(DocumentStorage &) override {}
  const VarnodeData &getRegister(const std::string &) const override {
    return dummy_register;
  }
  std::string getRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  std::string getExactRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  void getAllRegisters(std::map<VarnodeData, std::string> &) const override {}
  void getUserOpNames(std::vector<std::string> &) const override {}
  int4 instructionLength(const Address &) const override { return 0; }
  int4 oneInstruction(PcodeEmit &, const Address &) const override { return 0; }
  int4 printAssembly(AssemblyEmit &, const Address &) const override { return 0; }
};

static void render_function(const CommentDatabaseInternal &db, const Address &fad) {
  bool first = true;
  CommentSet::const_iterator iter = db.beginComment(fad);
  CommentSet::const_iterator enditer = db.endComment(fad);
  std::cout << '[';
  for (; iter != enditer; ++iter) {
    const Comment *comment = *iter;
    if (!first)
      std::cout << ';';
    first = false;
    std::cout << comment->getType() << ':'
              << comment->getFuncAddr().getOffset() << ':'
              << comment->getAddr().getOffset() << ':'
              << comment->getUniq() << ':' << comment->getText();
  }
  std::cout << ']';
}

static void run_roundtrip(FixtureTranslate &translate) {
  AddrSpace *ram = translate.getSpace(3);
  Address fad(ram, 0x1000);
  CommentDatabaseInternal source;
  source.addComment(Comment::header, fad, fad, "Header");
  source.addComment(Comment::warning, fad, Address(ram, 0x2000), "Test warning");
  source.addComment(Comment::warning, fad, Address(ram, 0x2000), "Second warning");
  bool duplicate = source.addCommentNoDuplicate(
      Comment::user2, fad, Address(ram, 0x2000), "Test warning");
  bool inserted = source.addCommentNoDuplicate(
      Comment::user2, fad, Address(ram, 0x2000), "User note");

  std::ostringstream encoded;
  XmlEncode encoder(encoded, false);
  source.encode(encoder);
  std::string xml = encoded.str();
  bool schema_ok = xml.find("<addr space=\"ram\" offset=\"0x1000\"/>") != std::string::npos &&
                   xml.find("<addr space=\"ram\" offset=\"0x2000\"/>") != std::string::npos &&
                   xml.find("<text>Test warning</text>") != std::string::npos;

  std::istringstream input(xml);
  XmlDecode decoder(&translate);
  decoder.ingestStream(input);
  CommentDatabaseInternal decoded;
  decoded.decode(decoder);

  std::cout << "case=roundtrip|schema=" << (schema_ok ? 1 : 0)
            << "|duplicate=" << (duplicate ? 1 : 0)
            << "|inserted=" << (inserted ? 1 : 0) << "|comments=";
  render_function(decoded, fad);
  std::cout << '\n';
}

static void run_filter(FixtureTranslate &translate) {
  AddrSpace *ram = translate.getSpace(3);
  Address fad(ram, 0x1000);
  Address other_fad(ram, 0x5000);
  CommentDatabaseInternal db;
  db.addComment(Comment::warningheader, fad, fad, "Warning header");
  db.addComment(Comment::header, fad, fad, "Header");
  db.addComment(Comment::warning, fad, Address(ram, 0x2000), "Inline warning");
  db.addComment(Comment::user2, fad, Address(ram, 0x3000), "User note");
  db.addComment(Comment::warning, other_fad, Address(ram, 0x6000), "Other warning");
  db.clearType(fad, Comment::warning | Comment::warningheader);

  std::cout << "case=filter|f1000=";
  render_function(db, fad);
  std::cout << "|f5000=";
  render_function(db, other_fad);
  std::cout << '\n';
}

static void run_unknown_type_partial(FixtureTranslate &translate) {
  AddrSpace *ram = translate.getSpace(3);
  Comment comment(Comment::header, Address(ram, 0xaaaa), Address(ram, 0xbbbb),
                  7, "sentinel");
  comment.setEmitted(true);
  std::string error;
  try {
    std::istringstream input("<comment type=\"bogus\"/>");
    XmlDecode decoder(&translate);
    decoder.ingestStream(input);
    comment.decode(decoder);
  }
  catch (const LowlevelError &err) {
    error = err.explain;
  }
  std::cout << "case=unknown_type|error=" << error
            << "|type=" << comment.getType()
            << "|emitted=" << (comment.isEmitted() ? 1 : 0)
            << "|func=" << comment.getFuncAddr().getOffset()
            << "|addr=" << comment.getAddr().getOffset()
            << "|uniq=" << comment.getUniq()
            << "|text=" << comment.getText() << '\n';
}

static void run_missing_offset_partial(FixtureTranslate &translate) {
  AddrSpace *ram = translate.getSpace(3);
  Comment comment(Comment::header, Address(ram, 0xaaaa), Address(ram, 0xbbbb),
                  7, "sentinel");
  comment.setEmitted(true);
  std::string error;
  try {
    std::istringstream input(
        "<comment type=\"warning\"><addr space=\"ram\"/>"
        "<addr space=\"ram\" offset=\"0xbbbb\"/><text>replacement</text></comment>");
    XmlDecode decoder(&translate);
    decoder.ingestStream(input);
    comment.decode(decoder);
  }
  catch (const LowlevelError &err) {
    error = err.explain;
  }
  std::cout << "case=missing_offset|error=" << error
            << "|type=" << comment.getType()
            << "|emitted=" << (comment.isEmitted() ? 1 : 0)
            << "|func=" << comment.getFuncAddr().getOffset()
            << "|addr=" << comment.getAddr().getOffset()
            << "|uniq=" << comment.getUniq()
            << "|text=" << comment.getText() << '\n';
}

static void run_unknown_property_encode(FixtureTranslate &translate) {
  AddrSpace *ram = translate.getSpace(3);
  Comment comment(64, Address(ram, 0xaaaa), Address(ram, 0xbbbb),
                  7, "sentinel");
  comment.setEmitted(true);
  std::ostringstream encoded;
  XmlEncode encoder(encoded, false);
  std::string error;
  try {
    comment.encode(encoder);
  }
  catch (const LowlevelError &err) {
    error = err.explain;
  }
  std::cout << "case=unknown_property_encode|error=" << error
            << "|stream_empty=" << (encoded.str().empty() ? 1 : 0)
            << "|type=" << comment.getType()
            << "|emitted=" << (comment.isEmitted() ? 1 : 0)
            << "|func=" << comment.getFuncAddr().getOffset()
            << "|addr=" << comment.getAddr().getOffset()
            << "|uniq=" << comment.getUniq()
            << "|text=" << comment.getText() << '\n';
}

int main() {
  AttributeId::initialize();
  ElementId::initialize();
  FixtureTranslate translate;
  std::cout << "schema=1|fixture=COMMENT-WARNING-CODEC-0001|oracle="
            << "e40ed13014025f82488b1f8f7bca566894ac376b\n";
  run_roundtrip(translate);
  run_filter(translate);
  run_unknown_type_partial(translate);
  run_missing_offset_partial(translate);
  run_unknown_property_encode(translate);
  return 0;
}

File: tests/oracle/comment_warning_codec_1204.rs
// COMMENT-WARNING-CODEC-0001: Rugra comparand for the locked Ghidra 12.0.4
// Comment and CommentDatabaseInternal codec/order/dedup/filter oracle.

use std::error::Error;
use std::sync::{Arc, RwLock};

use rugra::address::Address;
use rugra::comment::{comment_type, Comment, CommentDatabaseInternal};
use rugra::marshal::{AttributeId, ElementId, Encoder, IdRegistry, TreeDecoder, TreeEncoder};
use rugra::space::{space_flags, AddrSpace, SpaceType};

fn ram_space() -> AddrSpace {
    AddrSpace::new_space(
        SpaceType::Processor,
        "ram",
        false,
        8,
        1,
        3,
        space_flags::HASPHYSICAL,
        0,
        0,
    )
}

fn registry() -> Arc<RwLock<IdRegistry>> {
    let registry = Arc::new(RwLock::new(IdRegistry::new()));
    {
        let mut ids = registry.write().unwrap();
        for name in ["type", "space", "offset", "XMLcontent"] {
            ids.register_attribute(name);
        }
        for name in ["comment", "commentdb", "addr", "text"] {
            ids.register_element(name);
        }
    }
    registry
}

fn render_function(db: &CommentDatabaseInternal, fad: Address) {
    print!("[");
    for (index, comment) in db.comments_for_function(fad).enumerate() {
        if index != 0 {
            print!(";");
        }
        print!(
            "{}:{}:{}:{}:{}",
            comment.get_type(),
            comment.get_func_addr().as_u64(),
            comment.get_addr().as_u64(),
            comment.get_uniq(),
            comment.get_text()
        );
    }
    print!("]");
}

fn encoded_schema_ok(root: &Arc<RwLock<rugra::marshal::Element>>) -> bool {
    let root = root.read().unwrap();
    let mut saw_1000 = false;
    let mut saw_2000 = false;
    let mut saw_warning_text = false;
    for comment in root.get_children() {
        let comment = comment.read().unwrap();
        for child in comment.get_children() {
            let child = child.read().unwrap();
            if child.get_name() == "addr" && child.get_attribute_value("space") == Some("ram") {
                saw_1000 |= child.get_attribute_value("offset") == Some("4096");
                saw_2000 |= child.get_attribute_value("offset") == Some("8192");
            }
            if child.get_name() == "text"
                && child.get_attribute_value("XMLcontent") == Some("Test warning")
            {
                saw_warning_text = true;
            }
        }
    }
    saw_1000 && saw_2000 && saw_warning_text
}

fn run_roundtrip(ram: &AddrSpace) -> Result<(), Box<dyn Error>> {
    let fad = Address::with_space(ram, 0x1000);
    let mut source = CommentDatabaseInternal::new();
    source.add_comment(comment_type::HEADER, fad, fad, "Header");
    source.add_comment(
        comment_type::WARNING,
        fad,
        Address::with_space(ram, 0x2000),
        "Test warning",
    );
    source.add_comment(
        comment_type::WARNING,
        fad,
        Address::with_space(ram, 0x2000),
        "Second warning",
    );
    let duplicate = source.add_comment_no_duplicate(
        comment_type::USER2,
        fad,
        Address::with_space(ram, 0x2000),
        "Test warning",
    );
    let inserted = source.add_comment_no_duplicate(
        comment_type::USER2,
        fad,
        Address::with_space(ram, 0x2000),
        "User note",
    );

    let ids = registry();
    let mut encoder = TreeEncoder::new(ids.clone());
    source.encode(&mut encoder)?;
    let document = encoder.into_document();
    let root = document.get_root().unwrap().clone();
    let schema_ok = encoded_schema_ok(&root);
    let mut decoder = TreeDecoder::new(root, ids);
    let mut decoded = CommentDatabaseInternal::new();
    decoded.decode(&mut decoder)?;

    print!(
        "case=roundtrip|schema={}|duplicate={}|inserted={}|comments=",
        if schema_ok { 1 } else { 0 },
        if duplicate { 1 } else { 0 },
        if inserted { 1 } else { 0 }
    );
    // The current Decoder trait does not carry an AddrSpaceManager. The
    // decoded projection is therefore the complete ordered offset state.
    render_function(&decoded, Address::new(0x1000));
    println!();
    Ok(())
}

fn run_filter(ram: &AddrSpace) {
    let fad = Address::with_space(ram, 0x1000);
    let other_fad = Address::with_space(ram, 0x5000);
    let mut db = CommentDatabaseInternal::new();
    db.add_comment(comment_type::WARNINGHEADER, fad, fad, "Warning header");
    db.add_comment(comment_type::HEADER, fad, fad, "Header");
    db.add_comment(
        comment_type::WARNING,
        fad,
        Address::with_space(ram, 0x2000),
        "Inline warning",
    );
    db.add_comment(
        comment_type::USER2,
        fad,
        Address::with_space(ram, 0x3000),
        "User note",
    );
    db.add_comment(
        comment_type::WARNING,
        other_fad,
        Address::with_space(ram, 0x6000),
        "Other warning",
    );
    db.clear_type(fad, comment_type::WARNING | comment_type::WARNINGHEADER);

    print!("case=filter|f1000=");
    render_function(&db, fad);
    print!("|f5000=");
    render_function(&db, other_fad);
    println!();
}

fn run_unknown_type_partial() {
    let ids = registry();
    let comment_element = ElementId::new("comment", 0);
    let mut encoder = TreeEncoder::new(ids.clone());
    encoder.open_element(&comment_element);
    encoder.write_string(&AttributeId::new("type", 0), "bogus");
    encoder.close_element(&comment_element);
    let root = encoder.into_document().get_root().unwrap().clone();

    let mut comment = Comment::new(
        comment_type::HEADER,
        Address::new(0xaaaa),
        Address::new(0xbbbb),
        7,
        "sentinel",
    );
    comment.set_emitted(true);
    let mut decoder = TreeDecoder::new(root, ids);
    let error = comment.decode(&mut decoder).unwrap_err().to_string();
    println!(
        "case=unknown_type|error={}|type={}|emitted={}|func={}|addr={}|uniq={}|text={}",
        error,
        comment.get_type(),
        if comment.is_emitted() { 1 } else { 0 },
        comment.get_func_addr().as_u64(),
        comment.get_addr().as_u64(),
        comment.get_uniq(),
        comment.get_text()
    );
}

fn run_missing_offset_partial() {
    let ids = registry();
    let comment_element = ElementId::new("comment", 0);
    let addr_element = ElementId::new("addr", 0);
    let mut encoder = TreeEncoder::new(ids.clone());
    encoder.open_element(&comment_element);
    encoder.write_string(&AttributeId::new("type", 0), "warning");
    encoder.open_element(&addr_element);
    encoder.write_string(&AttributeId::new("space", 0), "ram");
    encoder.close_element(&addr_element);
    encoder.close_element(&comment_element);
    let root = encoder.into_document().get_root().unwrap().clone();

    let mut comment = Comment::new(
        comment_type::HEADER,
        Address::new(0xaaaa),
        Address::new(0xbbbb),
        7,
        "sentinel",
    );
    comment.set_emitted(true);
    let mut decoder = TreeDecoder::new(root, ids);
    let error = comment.decode(&mut decoder).unwrap_err().to_string();
    println!(
        "case=missing_offset|error={}|type={}|emitted={}|func={}|addr={}|uniq={}|text={}",
        error,
        comment.get_type(),
        if comment.is_emitted() { 1 } else { 0 },
        comment.get_func_addr().as_u64(),
        comment.get_addr().as_u64(),
        comment.get_uniq(),
        comment.get_text()
    );
}

fn run_unknown_property_encode(ram: &AddrSpace) {
    let mut comment = Comment::new(
        64,
        Address::with_space(ram, 0xaaaa),
        Address::with_space(ram, 0xbbbb),
        7,
        "sentinel",
    );
    comment.set_emitted(true);
    let mut encoder = TreeEncoder::new(registry());
    let error = comment.encode(&mut encoder).unwrap_err().to_string();
    println!(
        "case=unknown_property_encode|error={}|stream_empty={}|type={}|emitted={}|func={}|addr={}|uniq={}|text={}",
        error,
        if encoder.root().is_none() { 1 } else { 0 },
        comment.get_type(),
        if comment.is_emitted() { 1 } else { 0 },
        comment.get_func_addr().as_u64(),
        comment.get_addr().as_u64(),
        comment.get_uniq(),
        comment.get_text()
    );
}

fn main() -> Result<(), Box<dyn Error>> {
    let ram = ram_space();
    println!(
        "schema=1|fixture=COMMENT-WARNING-CODEC-0001|oracle=\
         e40ed13014025f82488b1f8f7bca566894ac376b"
    );
    run_roundtrip(&ram)?;
    run_filter(&ram);
    run_unknown_type_partial();
    run_missing_offset_partial();
    run_unknown_property_encode(&ram);
    Ok(())
}
